pkg/autoupdate/handler.go

```diff
@@ -1,52 +1,115 @@
-			level.Info(u.logger).Log("msg", "download failed", "target", u.target, "err", err)
+			level.Info(u.logger).Log(
+				"msg", "tuf updater returned",
+				"target", u.target,
+				"err", err)
 			return
 		}
 
-		if err := fs.UntarBundle(stagingPath, stagingPath); err != nil {
-			level.Info(u.logger).Log("msg", "untar downloaded target", "binary", u.target, "err", err)
+		level.Debug(u.logger).Log(
+			"msg", "Starting to handle a staged TUF download",
+			"file", stagingPath,
+			"target", u.target,
+		)
+
+		// We store the updated file in a dated directory. The
+		// dated directory is a bit odd, but it's plastering
+		// over how tuf works. This way we ensure we're always
+		// running the mostly recently downloaded file.  There
+		// are other patterns we should investigate if we
+		// change the way we denote stable in notary.
+		updateDir := filepath.Join(u.updatesDirectory, strconv.FormatInt(time.Now().Unix(), 10))
+
+		// Note that this is expecting the binary in the
+		// tarball to be named binaryName. There some some
+		// extension weirdness issues on windows vs posix.
+		outputBinary := filepath.Join(updateDir, u.binaryName)
+
+		if err := os.MkdirAll(updateDir, 0755); err != nil {
+			level.Error(u.logger).Log(
+				"msg", "making updated directory",
+				"dir", updateDir,
+				"err", err)
 			return
 		}
 
-		binary := filepath.Join(filepath.Dir(stagingPath), filepath.Base(u.destination))
-		if err := os.Rename(binary, u.destination); err != nil {
-			level.Info(u.logger).Log("msg", "update binary from staging dir", "binary", u.destination, "err", err)
+		// The UntarBundle(destination, source) paths are a
+		// little weird. Source is a tarball, obvious
+		// enough. But destination is a string that's passed
+		// through filepath.Dir. Which means it strips off the
+		// last component.
+		if err := fs.UntarBundle(outputBinary, stagingPath); err != nil {
+			level.Error(u.logger).Log(
+				"msg", "untar downloaded target",
+				"binary", outputBinary,
+				"err", err,
+			)
 			return
 		}
 
-		if err := os.Chmod(u.destination, 0755); err != nil {
-			level.Info(u.logger).Log("msg", "setting +x permissions on binary", "binary", u.destination, "err", err)
+		// Ensure it's executable
+		if err := os.Chmod(outputBinary, 0755); err != nil {
+			level.Error(u.logger).Log(
+				"msg", "setting +x permissions on binary",
+				"binary", outputBinary,
+				"err", err,
+			)
 			return
 		}
 
+		level.Info(u.logger).Log(
+			"msg", "Updated Binary ready to go",
+			"target", u.target,
+			"outputBinary", outputBinary,
+		)
+
 		if err := u.finalizer(); err != nil {
-			level.Info(u.logger).Log("msg", "calling restart function for updated binary", "binary", u.destination, "err", err)
+			// Some kinds of updates, require a full
+			// launcher restart. For example, windows
+			// doesn't have an exec, so launcher exits so
+			// the service manager will restart it. There
+			// may be others.
+			if IsLauncherRestartNeededErr(err) {
```

Comment:
could this be done in the finalizer method? Perhaps there's value in having it outside the finalizer too, not sure.

Reply:
I think moving it into the finalizer would be more thinking. I want to kick that forward.